File: agent/node/scope.ts
import type { Mapping } from "./mapping.ts";

export class ScopeError extends Error {}

export class ScopeGuard {
  readonly scope: string;
  private ids: Set<string>;

  constructor(mapping: Mapping) {
    this.scope = mapping.scope;
    this.ids = new Set(mapping.channels.map((c) => c.id));
  }

  allow(channelId: string): boolean {
    return this.ids.has(channelId);
  }

  redact(channelId: string): string {
    return this.allow(channelId) ? channelId : "<outside-scope>";
  }

  require(channelId: string, action: string): void {
    if (!this.allow(channelId)) {
      throw new ScopeError(`blocked ${action}: outside scope ${this.scope}`);
    }
  }

  drop(_channelId: string, reason: string): void {
    console.info(`dropped event source=<outside-scope> reason=${reason}`);
  }
}

File: agent/python/scope.py
"""Scope is a wall. Adapters never see ids outside the mapping."""

from __future__ import annotations

import logging

from mapping import Mapping


class ScopeError(PermissionError):
    pass


class ScopeGuard:
    def __init__(self, mapping: Mapping) -> None:
        self.scope = mapping.scope
        self._ids = mapping.ids()

    def allow(self, channel_id: str) -> bool:
        return channel_id in self._ids

    def redact(self, channel_id: str) -> str:
        return channel_id if self.allow(channel_id) else "<outside-scope>"

    def require(self, channel_id: str, action: str) -> None:
        if not self.allow(channel_id):
            raise ScopeError(
                f"blocked {action}: outside scope {self.scope!r}"
            )

    def drop(self, channel_id: str, reason: str) -> None:
        # The raw id never leaves this function.
        logging.info("dropped event source=<outside-scope> reason=%s", reason)
